cylc/flow/async_util.py

```diff
@@ -0,0 +1,340 @@
+        return self
+
+    def rewind(self):
+        """Return the head of the pipe."""
+        ptr = self
+        while ptr._left:
+            ptr = ptr._left
+        return ptr
+
+    def fastforward(self):
+        """Return the tail of the pipe."""
+        ptr = self
+        while ptr._right:
+            ptr = ptr._right
+        return ptr
+
+    def __iter__(self):
+        ptr = self.rewind()
+        while ptr._right:
+            yield ptr
+            ptr = ptr._right
+        yield ptr
+
+    def __repr__(self):
+        return ' | '.join((str(ptr) for ptr in self))
+
+    def __str__(self):
+        args = ''
+        if self.args:
+            args = ', '.join(map(repr, self.args))
+        if self.kwargs:
+            if args:
+                args += ', '
+            args += ', '.join(f'{k}={repr(v)}' for k, v in self.kwargs.items())
+        return f'{self.func.__name__}({args})'
+
+
+class _PipeFunction:
+    """Represent a function for use in an async pipe.
+
+    This class is just for syntactic sugar, it enables us to assign arguments
+    via the __call__ interface and enables us to add an interface for
+    preprocessing args.
+
+    """
+
+    def __init__(self, func, preproc=None):
+        self.func = func
+        self.preproc = preproc
+
+    def __call__(self, *args, filter_stop=True, **kwargs):
+        # assign args/kwargs to a function in a pipe
+        if self.preproc:
+            args, kwargs = self.preproc(*args, **kwargs)
+        return _AsyncPipe(
+            self.func,
+            args,
+            kwargs,
+            filter_stop
+        )
+
+    def __or__(self, other):
+        this = _AsyncPipe(self.func)
+        return this | other
+
+    async def __aiter__(self):
+        # this permits pipes with only one step
+        async for item in _AsyncPipe(self.func):
+            yield item
+
+    def __str__(self):
+        return _AsyncPipe(self.func).__str__()
+
+    def __repr__(self):
+        return _AsyncPipe(self.func).__repr__()
+
+    @property
+    def __doc__(self):
+        return self.func.__doc__
+
+
+def pipe(func=None, preproc=None):
+    """An asynchronous pipe implementation in pure Python.
+
+    Use this to decorate async functions in order to arrange them into
+    asynchronous pipes. These pipes can process multiple items through multiple
+    stages of the pipe simultaneously by doing what processing it can whilst
+    waiting on IO to take place in the background.
+
+    Args:
+        func (callable):
+            The function this decorator decorates.
+        preproc (callable):
+            An optional function for pre-processing any args or kwargs
+            provided to a function when the pipe is created.
+
+            preproc(args: tuple, kwargs: dict) -> (args: tuple, kwargs: dict)
+
+    Example:
+        A generator to begin our pipe with:
+        >>> @pipe
+        ... async def arange():
+        ...    for i in range(10):
+        ...        yield i
```

Comment:
> In your example the generator (speak) is not asynchronous as there is no await in it anywhere. 

I was following this example of yours in the code (although I admit to being confused about whether an async generator needs an await or not - I thought maybe `yield` passes control back to the event loop).

Reply:
(OK I think I understand how that works now).

Reply:
`yield` yields control to a generator in synchronous code, `await` yields control to a coroutine in asynchronous code. We have an async-generator here, both paradigms still apply but only in their own context.